Tasks/ServiceFabricDeploy/Create-DiffPackage.psm1

```diff
@@ -0,0 +1,149 @@
+
+        $app = Get-ServiceFabricApplication -ApplicationName $ApplicationName
+
+        # If $app is null, it means the application does not exist in the cluster. Diff Package is equal to Full Package. Use Full Package to do deployment
+        if (!$app -or $app.ApplicationTypeName -ne $applicationTypeName)
+        {
+            Write-Host (Get-VstsLocString -Key DIFFPKG_ApplicationDoesNotExist -ArgumentList @($ApplicationName, $ConnectedServiceEndpoint.Url))
+            Return
+        }
+
+        $diffPackagePath = Join-Path (Get-VstsInput -Name diffPackagePath -Require) "DiffPackage"
+        if (Test-Path -PathType Container -Path $diffPackagePath)
+        {
+            Remove-Item -Path $diffPackagePath -Recurse -Force
+        }
+        $diffPackagePath = New-Item -ItemType Directory -Path $diffPackagePath -Force
+        $diffPkgAppManifestPath = Join-Path $diffPackagePath $appManifestName
+
+        $clusterAppTypeVersion = $app.ApplicationTypeVersion
+
+        # If the ApplicationTypeVersion of the Application is not upgraded, no diff package is made because there is no need
+        if ($clusterAppTypeVersion -eq $localAppTypeVersion)
+        {
+            Write-Host (Get-VstsLocString -Key DIFFPKG_ApplicationIsNotChanged -ArgumentList @($ApplicationName, $clusterAppTypeVersion, $ConnectedServiceEndpoint.Url))
+            Return
+        }
+
+        # Get the service types from the cluster
+        $serviceTypes = Get-ServiceFabricServiceType -ApplicationTypeName $applicationTypeName -ApplicationTypeVersion $clusterAppTypeVersion
+        # Pack the service manifest names into an array
+        $clusterServiceManifestNames = $serviceTypes.ServiceManifestName
+
+        # If $clusterServiceManifestNames is null, it means no service types are registered. Diff Package is equal to Full Package. Use Full Package to do deployment
+        if (!$clusterServiceManifestNames)
+        {
+            Write-Host (Get-VstsLocString -Key DIFFPKG_NoServicesRunning -ArgumentList @($ApplicationName, $ConnectedServiceEndpoint.Url))
+            Return
+        }
+
+        Write-Host (Get-VstsLocString -Key DIFFPKG_CopyingToDiffPackge -ArgumentList @($localAppManifestPath, $diffPkgAppManifestPath))
+        Copy-Item $localAppManifestPath $diffPkgAppManifestPath -Force
+
+        # Get the service manifests from the cluster
+        $clusterServiceManifestByName = @{}
+        foreach ($clusterServiceManifestName in $clusterServiceManifestNames)
+        {
+            $clusterServiceManifestContent = Get-ServiceFabricServiceManifest -ApplicationTypeName $applicationTypeName -ApplicationTypeVersion $clusterAppTypeVersion -ServiceManifestName $clusterServiceManifestName
+            $clusterServiceManifestByName[$clusterServiceManifestName] = [XML]$clusterServiceManifestContent
+        }
+
+        foreach ($serviceManifestImport in $localAppManifestXml.ApplicationManifest.ServiceManifestImport)
```

Comment:
Can there be a scenario where locally a service is deleted? How do we update that in Diff method

Reply:
@sachinma In that case the ApplicationManifest won't have the deleted service name so the deleted service data won't be put into diff package.